### PR TITLE
Adds PDA chatroom app

### PR DESCRIPTION
**Pull request by Tastyfish**
I actually wanted to add this before, but then there was a suggestion thread, so I implemented it.
- You can have public channels everyone can see and join!
- You can have private channels that are invite-only!
 - Command can spy on either one via the messaging server terminal thing, so be careful what you say.
- Channels can have topics.
- People who have joined the channel (you can permanently leave them) will be notified of posts or other events via notifications in chat, like with Messenger messages. You can turn this feature off if people are being spammy.
 - You are also notified of channel invites.
- You can choose to make the chat UI automatically scroll to new messages or not.
 - This feature was also added to Messenger.
- Refactored PDA notifications so that any app can blink the PDA / play ringtones / etc.
- App icons now blink in the main menu when they notify.

### Screenshots
Imgur
Imgur
Imgur

### Changelog
:cl:
rscadd: Adds PDA chatrooms. Post memes! Talk to your fellow collaborators! Get caught!
tweak: You can now set the PDA messenger to automatically scroll down to new messages.
/:cl:

### CODE REVIEW

**Comment by Spacemanspark** [comment]
Yes.
:+1:
Next we need PDA cartridge games.

**Comment by johannhawk** [comment]
Sweet, now i can try to get all the staff from one department on one chatroom.

or "squads" of sec members in seperate and a unified chatrom.

**Comment by ProperPants** [comment]
> Sweet, now i can try to get all the staff from one department on one chatroom.
Departmental comms channels? I still like it though. :+1:

File: code/modules/pda/chatroom.dm [hunk 1]
```diff
@@ -1,16 +1,252 @@
...
-	var/list/muted = list()
-	var/topic = "" // topic message for the chatroom
-	var/password = "" // blank for no password.
-	var/operator = "" // name of the operator
-
-/datum/chatroom/proc/attempt_connect(var/obj/item/device/pda/device, var/obj/password)
-	if(!device)
+	var/topic = "Discussion" // topic message for the chatroom
+	var/is_public = 1
+	var/announcer = "CyberiadAI"
```
Comment (tigercat2000) [inline code/modules/pda/chatroom.dm]: how cute

File: code/modules/pda/chatroom.dm [hunk 2]
```diff
@@ -1,16 +1,252 @@
-var/list/chatrooms = list()
+var/list/chatrooms = list(new /datum/chatroom("General Discussion"))
 
 /datum/chatroom
```
Comment (tigercat2000) [inline code/modules/pda/chatroom.dm]: wouldn't this work better as a /datum/data/pda/chatroom?

**Comment by NTSAM** [comment]
lmao.

:+1: now i can be a piece of shit in chats ICly

File: code/modules/pda/chatroom.dm (hunk 2 above)
Comment (Tastyfish) [inline code/modules/pda/chatroom.dm]: That's the app's type, `/datum/chatroom` is actually the chatroom/channel people connect to.

Comment (tigercat2000) [inline code/modules/pda/chatroom.dm]: does the /datum/data/pda class have any inheritance stuff?

**Comment by KasparoVy** [comment]
lmao, I can see this going well..

File: nano/templates/pda_chatroom.tmpl [hunk 3]
```diff
@@ -0,0 +1,84 @@
+<!--
+Title: PDA Chatroom UI
+Used In File(s): \code\game\objects\items\devices\PDA\chatroom.dm
+ -->
+<div id="chatroom-header" class="item">
+	<div class="itemLabelNarrow">
+		<b>Chat Functions</b>:
+	</div>
+	<div class ="itemContent">
+		{{:helper.link(data.silent==1 ? 'Ringer: Off' : 'Ringer: On', data.silent==1 ? 'volume-off' : 'volume-up', {'choice' : "Toggle Ringer"}, null, 'pdalink fixedLeftWide')}}
```
Comment (tigercat2000) [inline nano/templates/pda_chatroom.tmpl]: spacing in these ==

File: nano/templates/pda_chatroom.tmpl [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		{{:helper.link(data.toff==1 ? 'Notifications: Off' : 'Notifications: On',data.toff==1 ? 'close' : 'check', {'choice' : "Toggle Chatroom"}, null, 'pdalink fixedLeftWide')}}
```
Comment (tigercat2000) [inline nano/templates/pda_chatroom.tmpl]: :eyes:

File: nano/templates/pda_chatroom.tmpl [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+		{{:helper.link('Set Ringtone', 'bell-o', {'choice' : "Ringtone"}, null, 'pdalink fixedLeftWide')}}
```
Comment (tigercat2000) [inline nano/templates/pda_chatroom.tmpl]: :eyes:

File: nano/templates/pda_chatroom.tmpl [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+						</div>
+						{{prevName = value.username;}}
+					{{/for}}
+				</div>
+				<div>
+					{{:helper.link('Post', 'comment', {'choice' : "Post"}, null, 'pdalink fixedLeftWidest')}}
+				</div>
+			</div>
+			<div id="userView" class="statusDisplay">
+				{{for data.users}}
```
Comment (tigercat2000) [inline nano/templates/pda_chatroom.tmpl]: what does this look like if you have a shitload of users

File: nano/templates/pda_chatroom.tmpl [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+					<div class="item">
+						<div class="itemLabel" style="width: 100%">{{:value}}</div>
+					</div>
+				{{/for}}
+			</div>
+			<script>
+				$(function() {
```
Comment (tigercat2000) [inline nano/templates/pda_chatroom.tmpl]: why is this invoking jquery

File: nano/templates/pda_messenger.tmpl [hunk 8]
```diff
@@ -29,7 +30,15 @@ Used In File(s): \code\game\objects\items\devices\PDA\core_apps.dm
 			</div>
 		</div>
 	</div>
-	{{:helper.link('Reply', 'comment', {'choice' : "Message", 'target': data.active_conversation}, null, 'pdalink fixedLeft')}}
+	{{:helper.link('Reply', 'comment', {'choice' : "Message", 'target': data.active_conversation}, null, 'pdalink fixedLeftWidest')}}
+	<script>
+		$(function() {
```
Comment (tigercat2000) [inline nano/templates/pda_messenger.tmpl]: same here

**Comment by TheDZD** [comment]
Does the message monitor console differentiate between different chat rooms?

**Comment by Fox-McCloud** [comment]
Just for clarity, you won't be in any channels by default, correct?

Spam is annoying enough, already--I'd really rather not have to change my PDA settings every single shift because some people want to talk about how <X> <Y> crewmember is for the 1000th time.

File: nano/templates/pda_chatroom.tmpl (hunk 6 above)
Comment (Tastyfish) [inline nano/templates/pda_chatroom.tmpl]: Exactly like the messenger user list

File: nano/templates/pda_chatroom.tmpl (hunk 7 above)
Comment (Tastyfish) [inline nano/templates/pda_chatroom.tmpl]: `$(function)` is shorthand for `$(document).ready(function)` which has no direct basic javascript equivalent.

**Comment by Tastyfish** [comment]
@TheDZD Yes, you view a list of chat rooms and select one.
@Fox-McCloud Yes, you are not in _any_ channels by default. You have to actively select even public channels, including the builting general chat one to actually join it and get messages.

File: nano/templates/pda_chatroom.tmpl (hunk 6 above)
Comment (Tastyfish) [inline nano/templates/pda_chatroom.tmpl]: The answer is "a tall ass list" which would only possibly look weird if you had more users than lines of posts, though if that's the case, then the players are doing something very strange indeed.

**Comment by Vivalas** [comment]
Nice, this has been on the minds of many a SS13 dev after stumbling across the unfinished code in the pda code.

This slightly nerfs tcomms blackouts and what not, might make antagging a bit difficult when everyone is in the same channel, plus security can meta to coordinate everything over a chatroom, although these aren't major things, wil be interesting to see how this affects antag/crew balance and gameplay in general.

**Comment by Tastyfish** [comment]
To clarify regarding blackouts, it does stop working if the messaging server goes down.

**Comment by tigercat2000** [comment]
@Tastyfish yeah, but, messaging server doesn't go down with tcomms, unfortunately

**Comment by joonasm** [comment]
will the pAI digital messenger be able to join chat rooms?

**Pull request merged**